### DISCUSSION TITLE
PiNN with measured inputs (U) and disturbances (d)

### DISCUSSION THREAD

**Discussion by mariaade26** [pnnl/neuromancer#225] [Q&A] [answered]
Hello everyone,
I am working on implementing a Physics-Informed Neural Network (PiNN) to learn the known ODE that describes building dynamics, along with its parameters (R, C). The ODE considers outdoor air temperature and solar radiation as disturbances, while HVAC thermal power serves as an input. These variables are represented as time-series data, available over a time span from tstart and tend. 
So far, I have tested the simplest case, where the ODE only includes the dependency on outdoor air temperature (T ext) and its corresponding data. I initially assumed that adding Text as an input to the network—alongside time—would help solve the problem. However, this approach does not seem to work. After 30,000 iterations, the results are highly inaccurate, as shown in the attached image. The estimated parameters exhibit significant errors: : R =  -2.4 and C = 2967, while the true ones are R = 0.005 and C = 5000. 
I would be really glad if someone would help me understand what am i doing wrong or how i can solve the problem. 

I would greatly appreciate any insights into what I might be doing wrong and how I can resolve this issue.
Thank you in advance for your help!
![Schermata del 2025-03-11 10-16-29](https://github.com/user-attachments/assets/d97813f0-cd8e-4903-acf9-302c8ad0d191)

I'm attaching here my code and data , hoping it would be helpful.

````
import torch
[ode_Text_dist.csv](https://github.com/user-attachments/files/19181056/ode_Text_dist.csv)

import torch.nn as nn                    
import numpy as np
from scipy.io import loadmat
import matplotlib.pyplot as plt
import warnings
import pandas as pd
from neuromancer.dataset import DictDataset
from neuromancer.modules import blocks
from neuromancer.system import Node
from neuromancer.constraint import variable
from neuromancer.loss import PenaltyLoss
from neuromancer.problem import Problem
from neuromancer.trainer import Trainer

warnings.filterwarnings("ignore")

torch.set_default_dtype(torch.float)
torch.manual_seed(1234)
np.random.seed(1234)

if torch.backends.mps.is_available():
    device = torch.device('mps')
elif torch.cuda.is_available():
    device = torch.device('cuda')
else:
    device = torch.device('cpu')


data = pd.read_csv('ode_Text_dist.csv')
ysol = data['0']
text = data['Text']
t = data['t']

T = torch.tensor(t.T).float()
Text = torch.tensor(text).float()
y_real = torch.tensor(ysol).float()

#test data
T_test = T.reshape(-1,1)
Text_test = Text.reshape(-1,1)
Y_test = y_real.reshape(-1,1) 

tempomin = T_test.min()
tempomax = T_test.max()

total_points = len(t)
Nf = 200
id_f = np.random.choice(total_points, Nf, replace=False)
T_train_Nu = T_test[id_f]
Y_train_Nu = Y_test[id_f]
Text_train_Nu = T_test[id_f]

Tmin = Text_train_Nu.min()
Tmax = Text_train_Nu.max()


T_train_Nu = (T_train_Nu -T_train_Nu.min())/(T_train_Nu.max() - T_train_Nu.min())
Text_train_Nu = (Text_train_Nu -Text_train_Nu.min()) / (Text_train_Nu.max()- Text_train_Nu.min())
T_test = (T_test - tempomin)/(tempomax - tempomin)
Text_test = (Text_test - Tmin)/(Tmax-Tmin)


T_train_Nu.requires_grad=True

train_data = DictDataset({'t':T_train_Nu, 'y':Y_train_Nu, 'text': Text_train_Nu}, name='train')
test_data = DictDataset({'t':T_test, 'y':Y_test, 'text':Text_test}, name='test')

batch_size = T_train_Nu.shape[0]  # full batch training
train_loader = torch.utils.data.DataLoader(train_data, batch_size=batch_size,
                                           collate_fn=train_data.collate_fn,
                                           shuffle=False)
test_loader = torch.utils.data.DataLoader(test_data, batch_size=batch_size,
                                         collate_fn=test_data.collate_fn,
                                         shuffle=False)


net = blocks.MLP(insize=2, outsize=1, hsizes=[32, 32], nonlin=nn.Tanh)
pde_net = Node(net, ['t', 'text'], ['y_hat'], name='net')
net_out = pde_net(train_data.datadict)


# symbolic Neuromancer variables
y = variable('y')           # PDE measurements from the dataset
y_hat = variable('y_hat')  # PDE solution generated as the output of a neural net (pde_net)
text = variable('text')
t = variable('t')          # temporal domain

R = variable(torch.nn.Parameter(torch.tensor(0.05)), display_name='R')       # trainable PDE parameter lambda
C = variable(torch.nn.Parameter(torch.tensor(4000.0)), display_name='C')             # trainable PDE parameter nu
# get the symbolic derivatives
dy_dt = y_hat.grad(t)
f_pinn = dy_dt - 1/(R*C)*(text - y_hat)

print(dy_dt({**net_out, **train_data.datadict}).shape)
print(f_pinn({**net_out, **train_data.datadict}).shape)
f_pinn.show()


scaling = 1000.
ell_1 = (f_pinn == 0.)^2
ell_2 = scaling*(y_hat == y)^2 
con_1 = 0
con_2 = 0


pinn_loss = PenaltyLoss(objectives=[ell_1, ell_2], constraints = [])
problem = Problem(nodes=[pde_net],     
                  loss=pinn_loss,       
                  grad_inference=True )


optimizer = torch.optim.AdamW(problem.parameters(), lr=0.001)
epochs = 30000

trainer = Trainer(
    problem.to(device),
    train_loader,
    optimizer=optimizer,
    epochs=epochs,
    epoch_verbose=200,
    train_metric='train_loss',
    dev_metric='train_loss',
    eval_metric="train_loss",
    warmup=epochs,
    device=device
)

best_model = trainer.train()
problem.load_state_dict(best_model)

print('True parameter R', 0.005)
print('Estimated parameter R = ', float(R.value))
print('True parameter C = ', 5000)
print('Estimated parameter C = ', float(C.value)
```)

**Comment by drgona** [comment, marked as answer]
Dear @mariaade26 thank you for using the library.

I looked through your setup and here are a few potential issues and suggestions that might help improve the performance of your PINN:
1. Input Normalization and Scaling
You're normalizing T_train_Nu and Text_train_Nu independently using their own min/max, but for PiNNs, inconsistent scaling between training and test sets can introduce instability. It's generally better to normalize using global min/max values from the entire dataset (training + testing), especially for variables that are time-correlated like Text.

2. In your code: `Text_train_Nu = T_test[id_f]  # Should be: Text_test[id_f]`
This looks like a bug. You're assigning temperature time values to the Text variable — this could be a key reason the model is failing to learn correctly.

3.  Loss Balancing. You're manually scaling the data fit term with:
```
scaling = 1000.
ell_2 = scaling * (y_hat == y)^2 
```
Try tuning this weight more systematically, ranging over different values, and analyze the convergence.

4. Make sure the input variable t has requires_grad=True. You set it for T_train_Nu, but when you use it in train_data, ensure it maintains gradient tracking — otherwise dy_dt might be zero or detached from the computation graph.

5. Before adding PINN loss, try training the network without physics constraints first (only on data fit) to ensure the architecture is at least capable of learning a decent approximation to the solution. This will ensure that the fitting part is bug free and you can focus on the PINN part.

Furthermore, I recommend going over [PINN examples ](https://github.com/pnnl/neuromancer?tab=readme-ov-file#physics-informed-neural-networks-pinns) in our library, which can provide more insights into successfully training PINNs.

Finally the problem you are trying to solve can be cast as system identification problem with hard constraints on the RC net model. For that we have specific example in [here](https://colab.research.google.com/github/pnnl/neuromancer/blob/master/examples/ODEs/Part_6_NetworkODE.ipynb). 

You could also use the RC constraints in learning neural state space models or neural ODEs. See associated examples [here ](https://colab.research.google.com/github/pnnl/neuromancer/blob/master/examples/domain_examples/NODE_building_dynamics.ipynb)and [here](https://colab.research.google.com/github/pnnl/neuromancer/blob/master/examples/domain_examples/NSSM_building_dynamics.ipynb).

Let me know if this helped :)

**Reply by mariaade26** [reply]
Hello @drgona,
first of all thank you very much for your answer and your suggestions!
The main issue of my code, as you noticed, was this error `Text_train_Nu = T_test[id_f]` (thanks again for pointing it out).
For the rest, as you suggested, I fixed the normalization handling and I made sure  that the network without physics constraints is suitable for my problem.

Right now, my model is capable of learning and approximated the solution to the ODE pretty well ( an example is in the attached graph), but I am still facing problems with the parameters' estimation. For identifiability reasons, I am trying to estimate the RC multiplication (that multiplies the time constant) and C,  that multiplies the other disturbances, properly scaled so that it has the same magnitude of RC.
`f_pinn = dy_dt - 1/(RC)*(text - y_hat)/1000 -1/C*(phi + phisun + gains)*1000/1000 `.

The estimation for the parameters stay quite  far away form the right value: actually, starting from the initial guess, the estimations move in the opposite direction, with respect to the right values (es. from 4, it reaches 2.5 instead of 5).

What do you think it can help fixing this issue? 
Thanks in advance!  :)
![complete_C_kJ](https://github.com/user-attachments/assets/ab27643e-80fc-455b-b839-58d436eb1bdc)

**Reply by drgona** [reply]
Hi @mariaade26, my apologies for the delayed response. 


What you’re running into looks like a classic non-identifiability issue: multiple parameter settings (and/or NN weights) produce essentially the same outputs and physics residuals. For thermal RC-type models with multiple disturbances (solar, ambient, internal gains), there are a few common invariances:
- Collinearity of inputs: solar, ambient, and HVAC inputs often move together in real data, so their effects can be swapped without changing the fit.
- NN–physics compensation: the network learns to absorb parameter errors, leaving the loss nearly unchanged along certain directions.

Some suggestions how to mitigate this:

**Fix one parameter; estimate the rest.** 
Breaking one degree of freedom collapses a large portion of the null space.

**Design inputs for persistency of excitation.**
Excite both fast and slow dynamics

**Neural net constraints**
add Lipschitz/monotonicity constraints

**Loss Weights **
The physics loss vs data‐fit loss balance might still be suboptimal. Even though the scaling is there, perhaps the weight is too low (or too high) relative to the magnitude of gradients from the data fit. It might help to sweep that weight or try adaptive weighting strategies (e.g. gradually increase physics constraint weight during training).

**Initialization Sensitivity**
The starting values for R and C might be in regions where the gradients lead away from the true values, especially if the loss surface has local minima or plateaus. You might try multiple random initializations, or even probing a grid for starting values to see how the outcomes differ.

**Simplifying the Physics Term** 
As a diagnostic, try using a simpler physics term (maybe only include one input at a time) and see if you can recover one parameter reliably before building up to full complexity.

**Regularization** 
Incorporating regularization, e.g., constraining parameters to plausible ranges, might help steer the estimation.

See attached [PINN training strategies.pdf](https://github.com/user-attachments/files/22432346/PINN.training.strategies.pdf)
